Private methods emit

mkubilayk opened the pull request:
This doesn't include private accessor, yet. Please ignore the fixture changes due to private accessors not being supported.

cc: @robpalme @dragomirtitian @molisani @joeywatts

mkubilayk committed:
Transform private methods

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Rename shouldTransformPrivateFields

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Accept baseline

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -898,21 +1009,57 @@ namespace ts {
+        }
+
+        function createHoistedVariableForPrivateName(privateName: string): Identifier {
+            const name = factory.createUniqueName("_" + privateName.substring(1), GeneratedIdentifierFlags.Optimistic);
```
This used to be `GeneratedIdentifierFlags.Optimistic | GeneratedIdentifierFlags.ReservedInNestedScopes` but `ReservedInNestedScopes` seems to be doing the opposite of what I would expect it to do. It caused shadowing generated variables in the upper scope. I will need to clarify this on the official PR.

mkubilayk replied:
Input: 
```ts
class Base {
    #x() { };
    constructor() {
        class Derived {
            #x() { };
            testBase(x: Base) {
                console.log(x.#x);
            }
        }
    }
}
```

Output (`GeneratedIdentifierFlags.Optimistic`):

```js
var __classPrivateMethodGet = (this && this.__classPrivateMethodGet) || function (receiver, accessCheck, fn) {
    if (!accessCheck.has(receiver)) {
        throw new TypeError("attempted to get private method on non-instance");
    }
    return fn;
};
var _x, _x_1;
class Base {
    constructor() {
        var _x_2, _x_3;
        _x.add(this);
        class Derived {
            constructor() {
                _x_2.add(this);
            }
            ;
            testBase(x) {
                console.log(__classPrivateMethodGet(x, _x_2, _x_3));
            }
        }
        _x_2 = new WeakSet(), _x_3 = function _x_3() { };
    }
    ;
}
_x = new WeakSet(), _x_1 = function _x_1() { };
```

Output (`GeneratedIdentifierFlags.Optimistic | GeneratedIdentifierFlags.ReservedInNestedScopes`):

```js
var __classPrivateMethodGet = (this && this.__classPrivateMethodGet) || function (receiver, accessCheck, fn) {
    if (!accessCheck.has(receiver)) {
        throw new TypeError("attempted to get private method on non-instance");
    }
    return fn;
};
var _x, _x_1;
class Base {
    constructor() {
        var _x_2, _x;
        _x.add(this);
        class Derived {
            constructor() {
                _x_2.add(this);
            }
            ;
            testBase(x) {
                console.log(__classPrivateMethodGet(x, _x_2, _x));
            }
        }
        _x_2 = new WeakSet(), _x = function _x() { };
    }
    ;
}
_x = new WeakSet(), _x_1 = function _x_1() { };
```

`_x_2` is correctly generated. But the next one reuses `_x` instead of `_x_3`, which is wrong.

joeywatts replied:
`ReservedInNestedScopes` seems to *restrict* the scope in which the name is reserved. (Names generated without that flag are guaranteed to not conflict with any other generated names.) I imagine you'd only want to use it for temporary variables with limited scope (say a loop counter for example, where you do not need it to be unique within the file). The name generation logic lives in emitter.ts, look for `reservedNamesStack` and `generatedNames`.

mkubilayk commented on src/compiler/factory/emitHelpers.ts:
```diff
@@ -34,6 +34,8 @@ namespace ts {
         createClassPrivateFieldGetHelper(receiver: Expression, privateField: Identifier): Expression;
         createClassPrivateFieldSetHelper(receiver: Expression, privateField: Identifier, value: Expression): Expression;
+        createClassPrivateMethodGetHelper(receiver: Expression, accessCheck: Identifier, fn: Identifier): Expression;
+        createClassPrivateReadonlyHelper(): Expression;
```
I'm planning to change the signature of this helper to `createClassPrivateReadonlyHelper(receiver: Expression, privateField: Identifier, value: Expression): Expression;`. This is so that we evaluate the value expression and be more spec-compliant on cases like [this](https://babeljs.io/en/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1XQMQCMwCcAUAlIgL4poBmA9pYYtAC4AWAlhAHRa7QC80wlAOwiUQAUzYhKAczwAiCI0oBXEABNokqQxYxyzHBHqyCAbmilSSAaIDucQmyo1TQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.12.12&externalPlugins=%40babel%2Fplugin-proposal-private-methods%407.12.1%2C%40babel%2Fplugin-proposal-class-properties%407.12.1). Untranspiled code would print the log first and then throw. Babel transform erases the log.

dragomirtitian commented:
@mkubilayk the test `privateNameDuplicateField.ts` seems to fail on master right now with a debug assertion 

```
 1) conformance tests
       conformance tests for tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts
         "before all" hook for "Correct errors for tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts":
     Error: Debug Failure. Unexpected private identifier placement
      at transformProperty (src\compiler\transformers\classFields.ts:834:47)
      at addPropertyStatements (src\compiler\transformers\classFields.ts:775:36)
```

mkubilayk commented:
> @mkubilayk the test `privateNameDuplicateField.ts` seems to fail on master right now with a debug assertion
> 
> ```
>  1) conformance tests
>        conformance tests for tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts
>          "before all" hook for "Correct errors for tests/cases/conformance/classes/members/privateNames/privateNameDuplicateField.ts":
>      Error: Debug Failure. Unexpected private identifier placement
>       at transformProperty (src\compiler\transformers\classFields.ts:834:47)
>       at addPropertyStatements (src\compiler\transformers\classFields.ts:775:36)
> ```

Yes - I'm aware. It will be fixed by private accessors implementation.

joeywatts reviewed (commented):
Can you try adding a test case like the following (which tests accessing a private method in a static property initializer)?

```typescript
class C {
    static instance: string = new C().#hello();
    #hello() { return "hello world"; }
}
```

This test will demonstrate that the function for the private method is defined and accessible when the static initializers for the class are evaluated.

mkubilayk committed:
Use a single WeakSet for brand-check

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Accept baseline

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -898,21 +1009,57 @@ namespace ts {
+
+                info = {
+                    placement: PrivateIdentifierPlacement.InstanceMethod,
+                    weakSetName: accessControlObject,
```
Good call - this was something we discussed with @dragomirtitian. I will make the change to use a single `WeakSet`.

mkubilayk committed:
Add a test for using private methods in static field initializers

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk commented:
@joeywatts @dragomirtitian pushed some changes:

- Add a test that uses private methods in public static field initializers
- Refactored the implementation to use a single `WeakSet` per-class
- Included class name as a prefix in generated variables to improve readability (not sure if there will be push-back on this because it changes private fields emit)

mkubilayk commented:
I should fix the `constructor` transform to do `_instances.add(this)` only once.

mkubilayk committed:
Add breaking checker test

Private methods inside class expressions should not error.

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Add to instances once per-instance

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Accept baseline

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
fix: evaluate receiver and rhs expressions before throwing on readonly assignment

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Add a test for evaluating rhs before readonly assignment

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Transpile private accessors

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
Accept baseline

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
fix: handle readonly/writeonly accessors

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
accept baseline

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
add a test for private setter without a getter

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
fix: getAllUnscopedEmitHelpers

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
fix: better handling of duplicate names

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

dragomirtitian commented on src/compiler/transformers/classFields.ts:
```diff
@@ -622,9 +775,14 @@ namespace ts {
                 properties = filter(properties, property => !!property.initializer || isPrivateIdentifier(property.name));
             }
 
+            const privateMethods = filter(
```
Is this really needed? `elements` already seems to contain all private methods props and accessors ..

mkubilayk replied:
Not sure what `elements` refer to here.

dragomirtitian replied:
Ah my bad. I was looking at part of the method above that has an elements.

dragomirtitian committed:
Fixed wrong error message for private methods in class expressions.

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -192,9 +203,49 @@ namespace ts {
+            Debug.assert(privateIdentifierInfo, "Undeclared private name for property declaration.");
+
+            if (privateIdentifierInfo.placement === PrivateIdentifierPlacement.InstanceMethod) {
+                return privateIdentifierInfo.functionName;
```
`privateIdentifierInfo.functionName` already contains the class name as a prefix. See:

https://github.com/bloomberg/TypeScript/pull/48/files#diff-2f470e1718434e5dfb841136a11eda3b4f46e9a2b3d14e0401c64f304da2b87eR1108-R1112

and

https://github.com/bloomberg/TypeScript/pull/48/files#diff-2f470e1718434e5dfb841136a11eda3b4f46e9a2b3d14e0401c64f304da2b87eR1176-R1186

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -923,6 +923,8 @@ namespace ts {
                                 privateIdentifierInfo.weakMapName
                             );
                         }
+                        default:
+                            return undefined;
```
Only when there are duplicate private names, which is an error and fails before emit.

dragomirtitian replied:
Does it still work if the compiler is set to emit on errors ?

mkubilayk replied:
That's the reason why I replaced the `Debug.fail` with `return undefined;`. It is still an open question whether we should be producing valid JS from invalid JS, which is what's happening in an existing test case:

https://github.com/microsoft/TypeScript/blob/master/tests/baselines/reference/privateNameDuplicateField.js

mkubilayk commented on src/compiler/factory/emitHelpers.ts:
```diff
@@ -844,6 +879,65 @@ namespace ts {
+        scoped: false,
+        text: `
+            var __classPrivateWriteonly = (this && this.__classPrivateWriteonly) || function () {
+                throw new TypeError("private element was defined without a getter");
```
It actually only applies to accessors. Do you think `private setter was defined without a getter` make more sense?

dragomirtitian replied:
yes I think that would be better.

mkubilayk committed:
change error message

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
add a test for async private methods with a higher target

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

joeywatts commented on src/compiler/transformers/classFields.ts:
```diff
@@ -9,20 +9,61 @@ namespace ts {
+    type PrivateIdentifierInfo =
+        | PrivateIdentifierInstanceField
+        | PrivateIdentifierInstanceMethod
+        | PrivateIdentifierInstanceGetterOnly
```
Would it be simpler to have a single `PrivateIdentifierInstanceAccessor` which has optional `getterName` and `setterName` properties?

mkubilayk replied:
I had that initially - it's pretty much the same complexity. This makes `addPrivateIdentifierToEnvironment` logic a little more complicated but then identifier access and assignment transforms can use `placement` in a unified way. If we had a single placement, `addPrivateIdentifierToEnvironment` would be a bit simpler but transforms would need to account of optionality.

joeywatts reviewed (commented):
For the accessors, can you make sure there is a test case that tests the return value of the setter expression? For example, `console.log(this.#myAccessorProp = 123)`. (I think the spec says the expression should always return the value being assigned - the RHS of the expression.)

mkubilayk commented on src/compiler/transformers/classFields.ts:
```diff
@@ -891,33 +1068,126 @@ namespace ts {
+                    const getterName = createHoistedVariableForPrivateName(text + "_get");
+
+                    if (previousInfo) {
+                        Debug.assert(previousInfo.placement === PrivateIdentifierPlacement.InstanceSetterOnly);
```
That's true - I guess we don't want to fail if the input is syntactically invalid?

joeywatts replied:
Yea, it should be impossible to fail these debug assertion statements with any input to the compiler. It's more of a check to fail loudly when the compiler hits a situation that's supposed to be impossible (which would be indicative of a bug in the compiler itself). Situations that can arise due to malformed programs should be flagged as errors during the parsing or checking phases, and then I believe the compiler always emits something if `noEmitOnError` is false, even if the code that it emits isn't valid due to syntax errors in the source code.

robpalme commented on src/compiler/transformers/classFields.ts:
```diff
@@ -810,6 +969,24 @@ namespace ts {
+        /**
+         * Generates brand-check initializer for private methods.
+         *
+         * @param methods An array of method declarations.
```
Missing param `statements`

mkubilayk committed:
fix: setter assignment returns rhs value

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
add a test for setter assignment return value

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
fix: handle duplicate accessors

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk committed:
add tests for duplicate accessors

Signed-off-by: Kubilay Kahveci <kahvecikubilay@gmail.com>

mkubilayk commented:
> For the accessors, can you make sure there is a test case that tests the return value of the setter expression? For example, `console.log(this.#myAccessorProp = 123)`. (I think the spec says the expression should always return the value being assigned - the RHS of the expression.)

Oh, good spot. Thanks! Fixed and added a test for this. Also changed the transform to handle duplicate accessors better.

robpalme reviewed (commented):
I was mid-review so just leaving these comments here.
